<root main_tree_to_execute="BehaviorTree">
    <!-- ////////// -->
    <BehaviorTree ID="BehaviorTree">
        <ReactiveFallback>
            <Inverter>
                <Condition ID="IsArretUrgence"/>
            </Inverter>
            <Condition ID="IsTirettePresente"/>
            <SequenceStar>
                <SubTree ID="initRobot"/>
                <ReactiveSequence>
                    <Action ID="AfficherLeScore"/>
                    <ReactiveFallback>
                        <SubTree ID="campJaune"/>
                        <SubTree ID="campViolet"/>
                    </ReactiveFallback>
                </ReactiveSequence>
            </SequenceStar>
        </ReactiveFallback>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="TesterToutesLesVentouses">
        <Fallback>
            <SubTree ID="VentouseD"/>
            <SubTree ID="VentouseCD"/>
            <SubTree ID="VentouseDG"/>
            <SubTree ID="VentouseC"/>
            <SubTree ID="VentouseCDG"/>
            <SubTree ID="VentouseG"/>
            <SubTree ID="VentouseCG"/>
        </Fallback>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="VentouseC">
        <Sequence>
            <Condition ID="IsVentouseCentre"/>
            <Inverter>
                <Condition ID="IsVentouseDroite"/>
            </Inverter>
            <Inverter>
                <Condition ID="IsVentouseGauche"/>
            </Inverter>
            <Action right_valve="false" left_valve="false" middle_valve="true" ID="Pince"/>
        </Sequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="VentouseCD">
        <Sequence>
            <Condition ID="IsVentouseCentre"/>
            <Condition ID="IsVentouseDroite"/>
            <Inverter>
                <Condition ID="IsVentouseGauche"/>
            </Inverter>
            <Action right_valve="true" left_valve="false" middle_valve="true" ID="Pince"/>
        </Sequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="VentouseCDG">
        <Sequence>
            <Condition ID="IsVentouseCentre"/>
            <Condition ID="IsVentouseDroite"/>
            <Condition ID="IsVentouseGauche"/>
            <Action right_valve="true" left_valve="true" middle_valve="true" ID="Pince"/>
        </Sequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="VentouseCG">
        <Sequence>
            <Condition ID="IsVentouseCentre"/>
            <Inverter>
                <Condition ID="IsVentouseDroite"/>
            </Inverter>
            <Condition ID="IsVentouseGauche"/>
            <Action right_valve="false" left_valve="true" middle_valve="true" ID="Pince"/>
        </Sequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="VentouseD">
        <Sequence>
            <Inverter>
                <Condition ID="IsVentouseCentre"/>
            </Inverter>
            <Condition ID="IsVentouseDroite"/>
            <Inverter>
                <Condition ID="IsVentouseGauche"/>
            </Inverter>
            <Action right_valve="true" left_valve="false" middle_valve="false" ID="Pince"/>
        </Sequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="VentouseDG">
        <Sequence>
            <Inverter>
                <Condition ID="IsVentouseCentre"/>
            </Inverter>
            <Condition ID="IsVentouseDroite"/>
            <Condition ID="IsVentouseGauche"/>
            <Action right_valve="true" left_valve="true" middle_valve="false" ID="Pince"/>
        </Sequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="VentouseG">
        <Sequence>
            <Inverter>
                <Condition ID="IsVentouseCentre"/>
            </Inverter>
            <Inverter>
                <Condition ID="IsVentouseDroite"/>
            </Inverter>
            <Condition ID="IsVentouseGauche"/>
            <Action right_valve="false" left_valve="true" middle_valve="false" ID="Pince"/>
        </Sequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="campJaune">
        <ReactiveSequence>
            <Inverter>
                <Condition ID="IsCampViolet"/>
            </Inverter>
            <Condition ID="MinuterieNonEcoulee"/>
            <SequenceStar>
                <SequenceStar>
                    <Action y="0" theta="0" ID="Aller" x="-0.5"/>
                    <Action y="0" theta="-1.57" ID="Aller" x="0.05"/>
                    <Action y="-1.275" theta="-1.57" ID="Aller" x="0.05"/>
                    <Action y="-1.18" theta="-1.57" ID="Aller" x="0.05"/>
                    <Action y="-1.18" theta="-1.57" ID="Aller" x="-0.5"/>
                </SequenceStar>
                <SequenceStar>
                    <Action y="0.0" theta="1.047" ID="Aller" x="-0.75"/>
                    <SubTree ID="debloquerGoldenium"/>
                    <Action y="0.75" theta="3.1416" ID="Aller" x="-0.70"/>
                    <SubTree ID="ramasserGoldenium"/>
                    <Action y="-0.15" theta="0" ID="Aller" x="0.50"/>
                    <SubTree ID="deposerGoldenium"/>
                </SequenceStar>
            </SequenceStar>
        </ReactiveSequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="campViolet">
        <ReactiveSequence>
            <Condition ID="IsCampViolet"/>
            <Condition ID="MinuterieNonEcoulee"/>
            <SequenceStar>
                <SequenceStar>
                    <Action y="0" theta="0" ID="Aller" x="-0.5"/>
                    <Action y="0" theta="1.57" ID="Aller" x="0.05"/>
                    <Action y="1.275" theta="1.57" ID="Aller" x="0.05"/>
                    <Action y="1.18" theta="1.57" ID="Aller" x="0.05"/>
                    <Action y="1.18" theta="1.57" ID="Aller" x="-0.5"/>
                </SequenceStar>
                <SequenceStar>
                    <Action y="0.0" theta="-1.047" ID="Aller" x="-0.75"/>
                    <SubTree ID="debloquerGoldenium"/>
                    <Action y="-0.75" theta="-3.1416" ID="Aller" x="-0.70"/>
                    <SubTree ID="ramasserGoldenium"/>
                    <Action y="0.15" theta="0" ID="Aller" x="0.50"/>
                    <SubTree ID="deposerGoldenium"/>
                </SequenceStar>
            </SequenceStar>
        </ReactiveSequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="debloquerGoldenium">
        <Action ms="10" ID="Attendre"/>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="deposerGoldenium">
        <Sequence>
            <Action angle_pince="140" angle_base="53" ID="Bras"/>
            <Action right_valve="false" left_valve="false" middle_valve="false" ID="Pince"/>
        </Sequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="initRobot">
        <Sequence>
            <Action score="50" ID="DefinirLeScore"/>
            <Action seconds="98" ID="LancerLaMinuterie"/>
        </Sequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="mettrePaletReserveArriere">
        <ReactiveSequence>
            <Inverter>
                <Condition ID="IsRobotEnMouvement"/>
            </Inverter>
            <ReactiveSequence>
                <Action angle_pince="54" angle_base="57" ID="Bras"/>
                <Action angle_pince="141" angle_base="66" ID="Bras"/>
                <Action right_valve="false" left_valve="false" middle_valve="false" ID="Pince"/>
            </ReactiveSequence>
        </ReactiveSequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="mettrePaletReserveAvant">
        <ReactiveSequence>
            <Inverter>
                <Condition ID="IsRobotEnMouvement"/>
            </Inverter>
            <ReactiveSequence>
                <Action angle_pince="54" angle_base="57" ID="Bras"/>
                <Action angle_pince="129" angle_base="63" ID="Bras"/>
                <Action right_valve="false" left_valve="false" middle_valve="false" ID="Pince"/>
            </ReactiveSequence>
        </ReactiveSequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="poserPaletAuSol">
        <ReactiveSequence>
            <Inverter>
                <Condition ID="IsRobotEnMouvement"/>
            </Inverter>
            <Action angle_pince="110" angle_base="70" ID="Bras"/>
            <Action angle_pince="110" angle_base="170" ID="Bras"/>
            <Action angle_pince="53" angle_base="175" ID="Bras"/>
            <Action angle_pince="23" angle_base="170" ID="Bras"/>
            <Action right_valve="false" left_valve="false" middle_valve="false" ID="Pince"/>
            <Action angle_pince="80" angle_base="170" ID="Bras"/>
            <Action angle_pince="53" angle_base="175" ID="Bras"/>
            <Action angle_pince="110" angle_base="71" ID="Bras"/>
        </ReactiveSequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="prendreEtPoserPaletAuSol">
        <Sequence>
            <SubTree ID="prendrePaletAuSol"/>
            <SubTree ID="poserPaletAuSol"/>
        </Sequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="prendrePaletAuSol">
        <ReactiveSequence>
            <Inverter>
                <Condition ID="IsRobotEnMouvement"/>
            </Inverter>
            <Fallback>
                <SequenceStar>
                    <Condition ID="IsPaletDroit"/>
                    <Condition ID="IsPaletGauche"/>
                    <Action angle_pince="80" angle_base="170" ID="Bras"/>
                    <Action angle_pince="51" angle_base="180" ID="Bras"/>
                    <Action angle_pince="27" angle_base="156" ID="Bras"/>
                    <SubTree ID="TesterToutesLesVentouses"/>
                    <Action angle_pince="110" angle_base="170" ID="Bras"/>
                    <Action angle_pince="110" angle_base="90" ID="Bras"/>
                </SequenceStar>
                <Sequence>
                    <Action angle_pince="110" angle_base="170" ID="Bras"/>
                    <Action angle_pince="110" angle_base="90" ID="Bras"/>
                </Sequence>
            </Fallback>
        </ReactiveSequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="prendrePaletReserve">
        <ReactiveSequence>
            <Inverter>
                <Condition ID="IsRobotEnMouvement"/>
            </Inverter>
            <ReactiveSequence>
                <Action angle_pince="21" angle_base="69" ID="Bras"/>
                <Action angle_pince="54" angle_base="111" ID="Bras"/>
                <Action angle_pince="54" angle_base="57" ID="Bras"/>
                <SubTree ID="TesterToutesLesVentouses"/>
                <Action angle_pince="110" angle_base="170" ID="Bras"/>
                <Action angle_pince="57" angle_base="54" ID="Bras"/>
            </ReactiveSequence>
        </ReactiveSequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="prendrePaletStockArriere">
        <ReactiveSequence>
            <Inverter>
                <Condition ID="IsRobotEnMouvement"/>
            </Inverter>
            <ReactiveFallback>
                <Action angle_pince="100" angle_base="36" ID="Bras"/>
                <Action angle_pince="110" angle_base="35" ID="Bras"/>
                <Action angle_pince="140" angle_base="35" ID="Bras"/>
                <Action angle_pince="150" angle_base="38" ID="Bras"/>
                <Action angle_pince="153" angle_base="36" ID="Bras"/>
                <SubTree ID="TesterToutesLesVentouses"/>
                <Action angle_pince="140" angle_base="40" ID="Bras"/>
                <Action angle_pince="90" angle_base="40" ID="Bras"/>
                <Action angle_pince="110" angle_base="70" ID="Bras"/>
            </ReactiveFallback>
        </ReactiveSequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="prendrePaletStockAvant">
        <ReactiveSequence>
            <Inverter>
                <Condition ID="IsRobotEnMouvement"/>
            </Inverter>
            <ReactiveFallback>
                <Action angle_pince="130" angle_base="36" ID="Bras"/>
                <Action angle_pince="140" angle_base="35" ID="Bras"/>
                <Action angle_pince="170" angle_base="35" ID="Bras"/>
                <Action angle_pince="180" angle_base="38" ID="Bras"/>
                <Action angle_pince="180" angle_base="32" ID="Bras"/>
                <SubTree ID="TesterToutesLesVentouses"/>
                <Action angle_pince="140" angle_base="40" ID="Bras"/>
                <Action angle_pince="90" angle_base="40" ID="Bras"/>
                <Action angle_pince="110" angle_base="70" ID="Bras"/>
            </ReactiveFallback>
        </ReactiveSequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="ramasserGoldenium">
        <ReactiveFallback>
            <Condition ID="IsRobotEnMouvement"/>
            <ReactiveSequence>
                <Inverter>
                    <Condition ID="IsCampViolet"/>
                </Inverter>
                <Action angle_pince="120" angle_base="156" ID="Bras"/>
                <SubTree ID="VentouseG"/>
                <Action angle_pince="110" angle_base="70" ID="Bras"/>
            </ReactiveSequence>
            <ReactiveSequence>
                <Condition ID="IsCampViolet"/>
                <Action angle_pince="120" angle_base="156" ID="Bras"/>
                <SubTree ID="VentouseD"/>
                <Action angle_pince="110" angle_base="70" ID="Bras"/>
            </ReactiveSequence>
        </ReactiveFallback>
    </BehaviorTree>
    <!-- ////////// -->
    <TreeNodesModel>
        <Action ID="AfficherLeScore"/>
        <Action ID="AjouterAuScore">
            <input_port name="score" default="1">Valeur à ajouter au score</input_port>
        </Action>
        <Action ID="Aller">
            <input_port name="theta" default="0">Angle du robot (en degree)</input_port>
            <input_port name="x" default="0">Position absolue en X (en metre)</input_port>
            <input_port name="y" default="0">Position absolue en Y (en metre)</input_port>
        </Action>
        <Action ID="Attendre">
            <input_port name="ms" default="1000">Temps en millisecondes</input_port>
        </Action>
        <Action ID="Bras">
            <input_port name="angle_base" default="0">Angle moteur au plus pres de la base</input_port>
            <input_port name="angle_pince" default="0">Angle moteur au plus pres de la pince</input_port>
        </Action>
        <Action ID="DefinirLeScore">
            <input_port name="score" default="0">Valeur du score</input_port>
        </Action>
        <Action ID="DefinirTempsMinuterie">
            <input_port name="seconds" default="1398">Temps de la minuterie en secondes</input_port>
        </Action>
        <Action ID="InitialiserPosition"/>
        <Condition ID="IsArretUrgence"/>
        <Condition ID="IsCampViolet"/>
        <Condition ID="IsPaletCentre"/>
        <Condition ID="IsPaletDroit"/>
        <Condition ID="IsPaletGauche"/>
        <Condition ID="IsRobotEnMouvement"/>
        <Condition ID="IsTirettePresente"/>
        <Condition ID="IsVentouseCentre"/>
        <Condition ID="IsVentouseDroite"/>
        <Condition ID="IsVentouseGauche"/>
        <Action ID="LancerLaMinuterie">
            <input_port name="seconds"/>
        </Action>
        <Condition ID="MinuterieNonEcoulee"/>
        <Action ID="Pince">
            <input_port name="left_valve" default="false"/>
            <input_port name="middle_valve" default="false"/>
            <input_port name="right_valve" default="false"/>
        </Action>
        <Condition ID="ScoreInferieurA">
            <input_port name="score" default="1">Score</input_port>
        </Condition>
        <Condition ID="ScoreSuperieurA">
            <input_port name="score" default="0">Score</input_port>
        </Condition>
        <SubTree ID="TestSensors"/>
        <SubTree ID="TesterToutesLesVentouses"/>
        <SubTree ID="VentouseC"/>
        <SubTree ID="VentouseCD"/>
        <SubTree ID="VentouseCDG"/>
        <SubTree ID="VentouseCG"/>
        <SubTree ID="VentouseD"/>
        <SubTree ID="VentouseDG"/>
        <SubTree ID="VentouseG"/>
        <Action ID="VitesseConstante">
            <input_port name="angular_z" default="0">Vitesse de rotation en rad/s</input_port>
            <input_port name="linear_x" default="0">Vitesse dans le sens avant/arriere en m/s</input_port>
            <input_port name="linear_y" default="0">Vitesse dans le sens droite/gauche en m/s</input_port>
        </Action>
        <SubTree ID="campJaune"/>
        <SubTree ID="campViolet"/>
        <Condition ID="checkPosition">
            <input_port name="x"/>
            <input_port name="y"/>
        </Condition>
        <SubTree ID="debloquerGoldenium"/>
        <SubTree ID="deposerGoldenium"/>
        <SubTree ID="initRobot"/>
        <SubTree ID="mettrePaletReserveArriere"/>
        <SubTree ID="mettrePaletReserveAvant"/>
        <SubTree ID="poserPaletAuSol"/>
        <SubTree ID="prendreEtPoserPaletAuSol"/>
        <SubTree ID="prendrePaletAuSol"/>
        <SubTree ID="prendrePaletReserve"/>
        <SubTree ID="prendrePaletStockArriere"/>
        <SubTree ID="prendrePaletStockAvant"/>
        <SubTree ID="ramasserGoldenium"/>
    </TreeNodesModel>
    <!-- ////////// -->
</root>

